### DISCUSSION TITLE
Curso Full Stack .Net - Dúvida sobre como lidar com consulta que tem dados de outro domínio

### DISCUSSION THREAD

**Discussion by suarezrafael** [balta-io/forum#639] [Tira Dúvidas] [answered]
Olá Balta. Talvez minha pergunta fuja do projeto aplicado porém fique a vontade para avaliar se cabe resposta.
Caso no projeto full stack eu avalie não usar um identity para gerenciar os acessos a api, portante não teria as tabelas IdentityUser entre outras.
Porém ainda sim quizesse armazenar o UserId na tabela Transaction 
Gostaria de complementar o projeto com uma consulta de transações, paginada e ordenada por nome e filtrada por nome (LIKE), que pudesse exibir o nome do usuário que efetuou as transações. Porém o nome do usuário estivesse armazenado em outra base de dados de outra api. 
Qual estratégia para esse caso poderia ser usado para evitar vazamento de dominio? Ou seja, não gravar o nome do usuário na api do controle financeiro, pois esse nome está em outra api ( provedor de identidade externo), mas mesmo assim, conseguir consultar na tabela Transaction com uma consulta paginada ordenando pelo nome do usuario que fez a transação. (ORDER BY external.UserName
OFFSET @offset ROWS FETCH NEXT @pageSize).

Isso seria um exemplo de quando algum requisito de consulta de dados é solicitado para a api fazer, porém ela não é detentora do dado usado para filtrar ou ordernar, ela só possui a FK do dado.

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @suarezrafael como vai?

Olá Balta. Talvez minha pergunta fuja do projeto aplicado porém fique a vontade para avaliar se cabe resposta.
- Fique à vontade!

Caso no projeto full stack eu avalie não usar um identity para gerenciar os acessos a api, portante não teria as tabelas IdentityUser entre outras.
- Plenamente possível

Porém ainda sim quizesse armazenar o UserId na tabela Transaction
- Como em um cenário com domínios separados e bancos separados por exemplo...

Gostaria de complementar o projeto com uma consulta de transações, paginada e ordenada por nome e filtrada por nome (LIKE), que pudesse exibir o nome do usuário que efetuou as transações. 
- Legal, se quiser montar a query e colocar aqui, eu avalio

Porém o nome do usuário estivesse armazenado em outra base de dados de outra api.
- Ai ou você duplica esta informação neste banco ou junta esta informação com as do usuário no APIM
- É a desvantagem dos microsserviços

Qual estratégia para esse caso poderia ser usado para evitar vazamento de dominio? 
- Uma mesma entidade pode estar presente em diferentes domínios, inclusive ter propriedades diferentes.
- Usuário pode existir em N domínios, e seus campos também, o que você precisa escolher é se vai duplicar estas informações (e depois manter isso atualizado) ou se vai recuperar as informações do usuário externamente a sua API

Ou seja, não gravar o nome do usuário na api do controle financeiro, pois esse nome está em outra api ( provedor de identidade externo), mas mesmo assim, conseguir consultar na tabela Transaction com uma consulta paginada ordenando pelo nome do usuario que fez a transação. (ORDER BY external.UserName
OFFSET @offset ROWS FETCH NEXT @pageSize).
- Normalmente os bancos vão ser separados também. Você não vai separar APIs e manter o mesmo banco para todas... uma modificação na API A pode quebrar o banco da API B
- Se quer manter tudo no mesmo banco, vai de monolítico
